#' @title Run rrBLUP polygenic inheritance on an R/qtl model
#'
#' @description
#' \code{polygenicQTL} Parses and pushes qtl model data int rrBLUP's mixed model
#' genomic selection pipe. Outputs polygenic inheritance stats.
#'
#' @param cross A qtl cross object containing conditional genotype probabilites calculated with
#' stepwidth = "fixed".
#' @param pheno.col Character or numeric vector of length 1,
#' indicating the phenotype to be tested.
#' @param qtl A QTL model, generated by  makeqtl (etc.) .
#' @param formula The formula for the qtl, names of columns in the covariate dataframe
#' and altnames of the QTL need to match exactly, or will be ignored.
#' @param covar the covariate specified in the model / formula. See fitqtl.
#' @param kinship.matrix A kinship matrix as supplied by rrBLUP::A.mat, or similar.
#' If not supplied, kinship matrix is calculated on the pseudomarker grid in the cross object
#' @param prob.thresh What should the probability of an allele need to be to be called?
#' @param rrBLUP.method should reml or ml be run?
#' @param verbose Should a normal or binary model be fit?
#' @param ... additional arguments passed on to rrBLUP::mixed.solve
#'
#' @details See rrBLUP::mixed.solve and rrBLUP::kin.blup
#'
#' @return A dataframe with 3 columns:
#' \itemize{
#'  \item{"Vg"}{The genetic variance attributable to the kinship (polygenic)}
#'  \item{"Ve"}{The amount of environmental (residual) varaince}
#'  \item{"polyH2"}{the proportion of Vtotal in Vg. Equivalent to the amount of
#'  heritability attributable to kinship (polygenic) variance following incoporpation of
#'  any fixed effects}
#' }
#' @examples
#' \dontrun{
#' library(qtlTools)
#' library(rrBLUP)
#' data(fake.bc)
#' cross<-fake.f2
#' cross<-calc.genoprob(cross, step = 12, stepwidth = "fixed")
#' qtl<-makeqtl(cross, chr = 13, pos = 28, what = "prob")
#' formula = " ~ Q1 + Q1*sex + sex"
#' covar = data.frame(sex = pull.pheno(cross, pheno.col = "sex"))
#' polygenicQTL(cross = cross, pheno.col = 1)
#' polygenicQTL(cross = cross, covar = covar, formula = "~ sex", pheno.col = 1)
#' polygenicQTL(cross = cross, qtl = qtl, formula = "~ Q1", pheno.col = 1)
#' polygenicQTL(cross = cross, qtl = qtl,covar = covar, formula = "~ Q1 + Q1*sex + sex", pheno.col = 1)
#' }
#'
#' @import qtl
#' @export
polygenicQTL<-function(cross, qtl = NULL, pheno.col = 1,
                       prob.thresh=0, covar=NULL,
                       formula = NULL,
                       rrBLUP.method = "REML",
                       kinship.matrix = NULL,
                       verbose = TRUE,
                       ...){

  if(!requireNamespace("rrBLUP", quietly = TRUE)){
    stop("install the rrBLUP package to run polygenicQTL\n")
  }else{
    requireNamespace("rrBLUP", quietly = TRUE)
  }

  if(is.null(qtl) & is.null(covar)){

    y = pull.pheno(cross, pheno.col=pheno.col)
    cross = subset(cross, ind = !is.na(y))
    covar<-covar[!is.na(y),]
    y<-pull.pheno(cross, pheno.col=pheno.col)

    if(is.null(kinship.matrix)){
      if(verbose) cat("calculating allele probabilities on a grid \n")
      gp<-genoprob2marker(cross, prob.thresh)

      if(verbose) cat("calculating kinship matrix \n")
      kinship.matrix<-rrBLUP::A.mat(gp)
    }

    if(verbose) cat("running polygenic selection \n")
    test<-rrBLUP::mixed.solve(y=y,
                              K = kinship.matrix,
                              method = rrBLUP.method, ...)
  }else{

    while(substr(formula,1,1) == " ")
      formula <- substr(formula,2,nchar(formula))
    if(substr(formula,1,1) == "y")
      formula <- substr(formula,2,nchar(formula))
    while(substr(formula,1,1) == " ")
      formula <- substr(formula,2,nchar(formula))

    if(is.null(covar)) covar = "1"
    if(is.null(qtl)) Q1 = "1"

    dat<-data.frame(covar,
                    genoprob2marker(cross,
                                    qtl = qtl,
                                    returnNumeric=FALSE,
                                    prob.thresh))

    y = pull.pheno(cross, pheno.col=pheno.col)
    is.good<-complete.cases(data.frame(dat,y))
    cross = subset(cross, ind = is.good)
    dat<-dat[is.good,]
    y<-pull.pheno(cross, pheno.col=pheno.col)

    formula = as.formula(formula)
    fixed.matrix<-model.matrix(formula, data = dat)

    if(is.null(kinship.matrix)){
      if(verbose) cat("calculating allele probabilities on a grid \n")
      gp<-genoprob2marker(cross, prob.thresh)

      if(verbose) cat("calculating kinship matrix \n")
      kinship.matrix<-rrBLUP::A.mat(gp)
    }

    if(verbose) cat("running polygenic selection \n")
    test<-rrBLUP::mixed.solve(y=y,
                              K = kinship.matrix,
                              X = fixed.matrix,
                              method = rrBLUP.method, ...)
  }
  return(data.frame(Vg = test$Vu,
                    Ve = test$Ve,
                    polyH2 = test$Vu/(test$Vu+test$Ve)))
}
